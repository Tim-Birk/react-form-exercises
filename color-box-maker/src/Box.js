import './Box.css';

const Box = ({ id, color, width, height, removeBox }) => {
  const style = {
    backgroundColor: color,
    width: `${width}px`,
    height: `${height}px`,
  };

  const handleRemove = (evt) => {
    removeBox(id);
  };

  const testId = `${color}-${width}-${height}`;
  return (
    <div className='Box'>
      <div data-testid={`Box-${testId}`} style={style}></div>
      <button data-testid={`Box-delete-${testId}`} onClick={handleRemove}>
        X
      </button>
    </div>
  );
};

export default Box;
