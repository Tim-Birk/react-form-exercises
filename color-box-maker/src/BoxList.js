import React, { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuid } from 'uuid';
import './BoxList.css';

const BoxList = () => {
  const [boxes, setBoxes] = useState([
    { id: uuid(), color: 'green', width: 100, height: 115 },
    { id: uuid(), color: 'yellow', width: 90, height: 85 },
    { id: uuid(), color: 'red', width: 190, height: 185 },
    { id: uuid(), color: 'black', width: 175, height: 140 },
  ]);

  const addBox = (color, width, height) => {
    const newBox = { id: uuid(), color, width, height };
    setBoxes((oldBoxes) => [...oldBoxes, newBox]);
  };

  const removeBox = (id) => {
    setBoxes(boxes.filter((box) => box.id !== id));
  };

  return (
    <>
      <h1>Box List</h1>
      <NewBoxForm addBox={addBox} />
      <div className='BoxList'>
        {boxes.map((box) => (
          <Box
            key={box.id}
            id={box.id}
            color={box.color}
            width={box.width}
            height={box.height}
            removeBox={removeBox}
          />
        ))}
      </div>
    </>
  );
};

export default BoxList;
