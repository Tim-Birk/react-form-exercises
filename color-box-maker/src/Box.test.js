import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Box from './Box';
import BoxList from './BoxList';

it('renders without crashing', function () {
  render(<Box />);
});

it('matches snapshot', function () {
  const { asFragment } = render(<Box />);
  expect(asFragment()).toMatchSnapshot();
});

it('can remove a new box', function () {
  const { getByLabelText, queryByText, queryByTestId } = render(<BoxList />);

  // no box yet
  expect(queryByTestId('Box-purple-120-100')).not.toBeInTheDocument();

  const colorInput = getByLabelText('Color:');
  const heightInput = getByLabelText('Height:');
  const widthInput = getByLabelText('Width:');
  const submitBtn = queryByText('Add New');

  // fill out the form
  fireEvent.change(colorInput, { target: { value: 'purple' } });
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(widthInput, { target: { value: 120 } });
  fireEvent.click(submitBtn);

  // box exists
  expect(queryByTestId('Box-purple-120-100')).toBeInTheDocument();
  const boxDeleteBtn = queryByTestId('Box-delete-purple-120-100');
  // box delete button exists
  expect(boxDeleteBtn).toBeInTheDocument();
  fireEvent.click(boxDeleteBtn);
  // box removed
  expect(queryByTestId('Box-purple-120-100')).not.toBeInTheDocument();
});
