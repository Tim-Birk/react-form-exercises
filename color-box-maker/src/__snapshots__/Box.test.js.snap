// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Box"
  >
    <div
      data-testid="Box-undefined-undefined-undefined"
    />
    <button
      data-testid="Box-delete-undefined-undefined-undefined"
    >
      X
    </button>
  </div>
</DocumentFragment>
`;
